### PR TITLE
Mchammer implementation

### PR DESCRIPTION
**Pull request by andrewtarzia**
Implementation of MCHammer optimisation in TopologyGraph.construct().

Questions we need to answer (some are commented too):
* Allow the user to specify the many settings as options (and how?) or just predefine some safe defaults (currently the case)?
* What about periodic systems? MCHammer does not modify the cell size.
* Handle output dir in MCHammer run - currently outputs trajectory and plots (could be optional)

### CODE REVIEW

File: src/stk/molecular/topology_graphs/topology_graph/topology_graph/topology_graph.py [hunk 1]
```diff
@@ -478,6 +489,79 @@ def _run_reactions(self, state):
...
+            step_size=0.25,
+            target_bond_length=1.2,
+            num_steps=1000,
+        )
+        mch_mol = optimizer.optimize(
+            mol=mch_mol,
+            bond_pair_ids=long_bond_ids,
+        )
+
+        return state.with_optimization_results(
```
Comment (andrewtarzia) [inline src/stk/molecular/topology_graphs/topology_graph/topology_graph/topology_graph.py:561]: change to with_position_matrix.

**Comment by andrewtarzia** [comment]
@lukasturcani have a look at the latest API whenever you can!

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+import mchammer as mch
+
+
+class Collapser(Optimizer):
+    """
+    Perform rigid-body collapse of molecule.
+
+    Implementation: [1]_
+
```
Comment (lukasturcani) [inline src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py:22]: Add references section https://numpydoc.readthedocs.io/en/latest/format.html

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    merge_subunits,
+    OptimizationIncompleteError,
+)
+
+import mchammer as mch
+
+
+class Collapser(Optimizer):
+    """
+    Perform rigid-body collapse of molecule.
```
Comment (lukasturcani) [inline src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py:19]:
```
Perform rigid-body collapse using MCHammer [1]_.
```

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+    Implementation: [1]_
+
+    .. [1] https://github.com/andrewtarzia/MCHammer.
```
Comment (lukasturcani) [inline src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py:23]: Remove full stop here.

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        save_trajectory=False,
+    ):
+        """
+        Initialize an instance of :class:`.Collapser`.
+
+        Parameters
+        ----------
+        step_size : :class:`float`, optional
+            The relative size of the step to take during collapse.
+            Defaults to 0.1 Angstrom.
```
Comment (lukasturcani) [inline src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py:41]: Do not duplicate defulat value in docstring.

Change to something like

```
The relative size of the step to take during collapse in Angstrom.
```

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+            threshold for halting collapse in Angstrom.
+            Defaults to 1.5 Angstrom.
+
+        scale_steps : :class:`bool`, optional
+            Whether to scale the step of each distict building block
+            by their relative distance from the molecules centroid.
+            Defaults to ``True``
+
+        save_trajectory : :class:`bool`, optional
+            `True` to save Collapser trajectory and information in
```
Comment (lukasturcani) [inline src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py:54]:
```
``True`` is for literal / fixed-width text, `True` is for interpreted text (ie text roles eg :class:`Something`)
```

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+            :attr:`_trajectory_data`.
```
Comment (lukasturcani) [inline src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py:55]: Expose `_trajecotry_data` in some sort of public api, like a getter

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+            self._trajectory_data = mch_result
+        else:
+            del mch_result
+            del data
+
+        return state.with_position_matrix(
+            mch_result.get_final_position_matrix()
+        )
+
+    def get_trajectory_results(self):
```
Comment (lukasturcani) [inline src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py:117]: get rid of this method and make an exmpale of using mchammer directly if you want trajectory

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/mchammer.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+            nonbond_sigma=nonbond_sigma,
+            nonbond_mu=nonbond_mu,
+            beta=beta,
+            random_seed=random_seed,
+        )
+        self._save_trajectory = save_trajectory
+        self._trajectory_data = None
+
+    def optimize(self, state):
+        """
```
Comment (lukasturcani) [inline src/stk/molecular/topology_graphs/topology_graph/optimizer/mchammer.py:109]: No need to repeat docstrings of an inhertied method, unless you are changing their content.

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/utilities.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+from mchammer import Bond
+
+
+class OptimizationIncompleteError(Exception):
+    ...
+
+
+def get_mch_bond_topology(state):
+    """
+    Returns bonds with atom1_id < atom2_id and long bonds to optimize.
```
Comment (lukasturcani) [inline src/stk/molecular/topology_graphs/topology_graph/optimizer/utilities.py:15]: dont underdstand

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/utilities.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
+"""
+This module defines utilities for optimizers.
+
+"""
+
+from mchammer import Bond
```
Comment (lukasturcani) [inline src/stk/molecular/topology_graphs/topology_graph/optimizer/utilities.py:6]: ideally dont import Bond directly

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/utilities.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    long_bond_ids = []
+    mch_bonds = []
+    for i, bond_infos in enumerate(state.get_bond_infos()):
+        ba1 = bond_infos.get_bond().get_atom1().get_id()
+        ba2 = bond_infos.get_bond().get_atom2().get_id()
+        # Must ensure bond atom id ordering is such that
+        # atom1_id < atom2_id.
+        if ba1 < ba2:
+            mch_bonds.append(
+                Bond(id=i, atom1_id=ba1, atom2_id=ba2)
```
Comment (lukasturcani) [inline src/stk/molecular/topology_graphs/topology_graph/optimizer/utilities.py:28]: Make MCHammer do the ordering, not the user of MCHammer

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/utilities.py [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+            long_bond_ids.append(ids)
+
+    return long_bond_ids, mch_bonds
+
+
+def merge_subunits(state, subunits):
+    """
+    Merge subunits in stk.Molecule by building block ids.
+
+    """
```
Comment (lukasturcani) [inline src/stk/molecular/topology_graphs/topology_graph/optimizer/utilities.py:51]: reduntant function, must use bulding block id directly

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py (hunk 2 above)
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py:22]: Done.

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py (hunk 3 above)
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py:19]: Done.

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py (hunk 4 above)
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py:23]: Done.

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py (hunk 5 above)
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py:41]: Done.

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py (hunk 6 above)
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py:54]: Done.

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py (hunk 7 above)
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py:55]: Done.

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py (hunk 8 above)
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/topology_graph/optimizer/collapser.py:117]: Done,

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/mchammer.py (hunk 9 above)
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/topology_graph/optimizer/mchammer.py:109]: Done.

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/utilities.py (hunk 10 above)
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/topology_graph/optimizer/utilities.py:15]: Done

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/utilities.py (hunk 11 above)
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/topology_graph/optimizer/utilities.py:6]: Done.

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/utilities.py (hunk 12 above)
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/topology_graph/optimizer/utilities.py:28]: Done.

File: src/stk/molecular/topology_graphs/topology_graph/optimizer/utilities.py (hunk 13 above)
Comment (andrewtarzia) [reply src/stk/molecular/topology_graphs/topology_graph/optimizer/utilities.py:51]: Done

**Comment by andrewtarzia** [comment]
Currently, none of these optimizers can modify the periodic cell. 

This should be a separate issue as it requires modifying the API of the PeriodicInfo class (from my experience).

**Comment by andrewtarzia** [comment]
@lukasturcani on review, please check the docs for each topology graph with a recommended optimizer -- what are your thoughts on this?

**Review by lukasturcani — APPROVED** [review]
Amazing, ty!

**Pull request merged**